src/main/java/fi/aalto/cs/apluscourses/intellij/model/APlusProject.java

```diff
@@ -29,7 +42,75 @@ public Path getBasePath() {
+  }
+
+  /**
+   * Creates a local course file containing the given URL. If a course file already exists (even if
+   * it was created with a different URL), this method does nothing and returns false.
+   * @param sourceUrl The URL that is added to the course file.
+   * @return True if the course file was created successfully, false if a course file already
+   *         exists.
+   * @throws IOException If an IO error occurs while creating the file.
+   */
+  public boolean createCourseFile(@NotNull URL sourceUrl) throws IOException {
+    File courseFile = getCourseFilePath().toFile();
+    if (courseFile.isFile()) {
+      return false;
+    }
+
+    JSONObject jsonObject = new JSONObject();
+    jsonObject.put("url", sourceUrl.toString());
+    FileUtils.writeStringToFile(courseFile, jsonObject.toString(), StandardCharsets.UTF_8);
+    return true;
+  }
+
+  /**
+   * Parses the stored URL from the course file and returns it, or null if the course file doesn't
+   * exist.
+   * @return The URL in the course file, or null if the course file doesn't exist.
+   * @throws IOException   If an IO error occurs while reading the course file.
+   * @throws JSONException If the JSON in the course file is malformed.
+   */
+  @Nullable
+  public URL getCourseFileUrl() throws IOException {
+    File courseFile = getCourseFilePath().toFile();
+    if (!courseFile.isFile()) {
+      return null;
+    }
+
+    // TODO: the course file is written as UTF-8, ensure that this always works
+    JSONTokener tokenizer = new JSONTokener(new FileInputStream(courseFile));
+    JSONObject jsonObject = new JSONObject(tokenizer);
+    return new URL(jsonObject.getString("url"));
+  }
+
+  private static final Object courseFileLock = new Object();
+  
+  /**
+   * Adds an entry for the given module to the course file. If an entry exists with the same name,
+   * it is overwritten.
+   * @param module The module for which an entry is added.
+   * @throws IOException   If an IO error occurs (for an example, the course file doesn't exist).
+   * @throws JSONException If the existing JSON in the course file is malformed.
+   */
+  public void addCourseFileEntry(@NotNull Module module) throws IOException {
+    File courseFile = getCourseFilePath().toFile();
+    synchronized (courseFileLock) {
+      JSONTokener tokenizer = new JSONTokener(new FileInputStream(courseFile));
+      JSONObject jsonObject = new JSONObject(tokenizer);
+
+      // It's possible that the "modules" key doesn't exist yet
+      JSONObject modulesObject = jsonObject.optJSONObject("modules");
+      if (modulesObject == null) {
+        modulesObject = new JSONObject();
+      }
+
+      JSONObject entry = new JSONObject().put("id", module.getVersionId());
+      modulesObject.put(module.getName(), entry);
+      jsonObject.put("modules", modulesObject);
+      FileUtils.writeStringToFile(courseFile, jsonObject.toString(), StandardCharsets.UTF_8);
+    }
```

Comment:
This is an interesting design choice. 🤔 And by no means I'd say it's bad. It's actually kinda cool but let's discuss it because it's something that I wouldn't come up with.

If I understand this correctly, the things that are stored in this JSON file are never kept in the main memory but instead, they are read from / written to the file each time they are accessed.

I'd find more conventional a solution where the JSON file is parsed to some domain-specific objects in startup and then those objects are modified (concurrency control happening there) and then the file is rewritten when some `save()` method is called. Did you consider this?

I can see there might be some advantages in your design but I'd really like to hear how you reasoned this out.

Reply:
hm... it seems to be the same place that attracted my attention as well (thus, in different PR)
https://github.com/Aalto-LeTech/intellij-plugin/pull/177#discussion_r427133533 :yum:

Reply:
A quick recap of what was discussed in the meeting for reference:

Reply:
A recap of what was discussed in the meeting for reference:
  + The state of the course file could be stored and maintained in `IntelliJCourse`
  + Either way, we still want to write to the file after each module import
    + In the distant future, we could consider "bulk writes" when multiple modules are imported at the same time
  + This stateless approach removes the possibility of the course file state in memory being inconsistent with the file on the disk
  + For now we'll go with this (assuming that synchronizing is sound, i.e. no deadlocks etc.) and refactor this later as needed